src/EditorFeatures/Core/BraceMatching/AbstractEmbeddedLanguageBraceMatcher.cs

```diff
@@ -2,44 +2,154 @@
-                    if (braceMatcher != null)
+                    foreach (var child in node.ChildNodesAndTokens())
                     {
-                        var result = await braceMatcher.FindBracesAsync(
-                            document, position, options, cancellationToken).ConfigureAwait(false);
-                        if (result != null)
+                        if (child.IsNode)
                         {
-                            return result;
+                            var value = Recurse(child.AsNode()!);
+                            if (value.HasValue)
+                                return value;
+                        }
+                        else
+                        {
+                            var value = ProcessToken(child.AsToken());
+                            if (value.HasValue)
+                                return value;
                         }
                     }
                 }
+
+                return null;
             }
 
-            return null;
+            private BraceMatchingResult? ProcessToken(SyntaxToken token)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+                return ClassifyToken(token);
+            }
+
+            private BraceMatchingResult? ClassifyToken(SyntaxToken token)
+            {
+                if (token.Span.IntersectsWith(_position) && _service.SyntaxTokenKinds.Contains(token.RawKind))
+                {
+                    _serviceBuffer.Clear();
+
+                    // First, see if this is a string annotated with either a comment or [StringSyntax] attribute. If
+                    // so, delegate to the first classifier we have registered for whatever language ID we find.
+                    if (_service.Detector.IsEmbeddedLanguageToken(token, _semanticModel, _cancellationToken, out var identifier, out _) &&
+                        _service.IdentifierToServices.TryGetValue(identifier, out var services))
+                    {
+                        foreach (var service in services)
+                        {
+                            // keep track of what classifiers we've run so we don't call into them multiple times.
+                            _serviceBuffer.Add(service.Value);
+
+                            // If this service added values then need to check the other ones.
+                            var result = service.Value.FindBraces(_semanticModel, token, _position, _options, _cancellationToken);
+                            if (result.HasValue)
+                                return result;
+                        }
+                    }
+
+                    // It wasn't an annotated API.  See if it's some legacy API our historical classifiers have direct
+                    // support for (for example, .net APIs prior to Net6).
+                    foreach (var legacyService in _service.LegacyServices)
```

Comment:
is it possible to keep these service maps separate in the `_service` itself?  That way we don't need to keep them separate in both the brace matcher here and in the classification one.

Reply:
unfortunately no. the idea of ths layer is that *we* (Roslyn) own finding the literals and determining what languages apply to them.  The only thing we're supporting from this point onwards is apis explicitly annotated with `[StringSyntax]`.  However, because we already shipped Regex/Json support, and we want it to work downlevel (e.g. on .net framework) we have this escape hatch to only let those two embedded languages attempt to classify any string literal using their own detection logic (which doesn't have to look for StringSyntaxAttribute).